### PR TITLE
Phase 5: Wall Distance Feature — Physics-Informed Input for Boundary Layer (8 parallel)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Add wall distance (minimum distance to nearest airfoil surface node) as a 25th input feature. Wall distance is a fundamental quantity in RANS turbulence modeling — it determines boundary layer thickness, wall shear stress, and the transition between near-wall and free-stream physics. The model currently has signed distance fields (SDF) but no explicit wall distance.

**Why this should work:**
- In Phase 5 Round 1, nezuko's walldist experiment (#1875) showed the walldist variant beat its own baseline by 6.5% (0.409 vs 0.442 val/loss). However, that baseline was missing pressure_first, so it wasn't compared to our full baseline.
- Wall distance is THE single most important feature in RANS turbulence modeling (used by k-omega SST, SA, etc.)
- The existing SDF features capture distance to the nearest surface with sign, but SDF is relative to the closest foil boundary. Wall distance captures the minimum distance to ANY surface node, which is different near the gap between foils.
- Partially redundant with existing dist_feat but provides a cleaner geometric signal.
- Zero architecture change. Just adds 1 extra input dimension.

**Previous result (PR #1875, nezuko):** Promising but incomplete — tested without pressure_first. This is the proper controlled test.

## Instructions

See first comment for implementation.

## Baseline (8-seed)
val/loss 0.404+/-0.004, p_in 13.33+/-0.58, p_oodc 8.37+/-0.22, p_tan 33.57+/-0.44, p_re 24.58+/-0.13

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Update — Walldist precomputation performance issue

The `torch.cdist` approach for computing wall distance is extremely slow when 6 GPU processes compete for CPU. After 60+ minutes, GPUs 0-5 (walldist experiments) are STILL precomputing wall distances and haven't started training.

**Root cause:** Computing `cdist(N_total, N_surf)` for each of ~1889 samples with N_total~85K is O(N_total × N_surf) per sample. With 6 concurrent processes saturating all CPU cores (~12000% total CPU), each process takes 60+ minutes.

**Baselines (GPUs 6-7) are training normally** at epoch 40+ and will complete on schedule.

**Proposed fix:** Replace `torch.cdist` with `scipy.spatial.cKDTree` which is O(N_total × log(N_surf)) — expected ~100x faster. This requires restarting the walldist experiments.

Awaiting advisor guidance: should I kill and restart with KDTree, or wait for cdist to finish?

Timestamp: 2026-03-29 ~00:00 UTC

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results

All 8 experiments completed successfully. Wall distance feature implemented using `scipy.spatial.cKDTree` for O(N·log(N_surf)) computation (after fixing initial `torch.cdist` performance issue — see earlier comment).

### Individual Results

| GPU | Config | Seed | Epochs | val/loss | p_in | p_tan | p_oodc | p_ood_re | W&B |
|-----|--------|------|--------|----------|------|-------|--------|----------|-----|
| 0 | walldist-log | 42 | 155 | **0.3896** | 13.2 | 33.8 | 8.8 | 24.6 | [4yjhkj0u](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/4yjhkj0u) |
| 1 | walldist-raw | 42 | 158 | 0.3947 | 13.8 | 34.4 | 8.5 | 24.5 | [00bqxjij](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/00bqxjij) |
| 2 | walldist-log | 43 | 158 | **0.3864** | **12.6** | 34.4 | 8.4 | **24.3** | [b8819tl5](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/b8819tl5) |
| 3 | walldist-log | 44 | 158 | 0.3919 | 13.1 | 33.9 | 8.7 | 24.8 | [n499st4x](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/n499st4x) |
| 4 | walldist-raw | 43 | 157 | 0.3954 | 13.2 | 34.3 | 9.0 | 25.0 | [mb017fys](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/mb017fys) |
| 5 | walldist-log lr=3e-4 | 42 | 158 | 0.3940 | 13.7 | 35.3 | 8.7 | 24.7 | [df23u1ip](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/df23u1ip) |
| 6 | baseline | 42 | 145 | 0.4042 | 13.3 | 33.5 | 8.3 | 24.9 | [8jthqm4k](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/8jthqm4k) |
| 7 | baseline | 43 | 145 | 0.4123 | 13.3 | 35.0 | 8.8 | 24.9 | [ou2hr1s7](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/ou2hr1s7) |

### Aggregated Means

| Config | Seeds | val/loss | p_in | p_tan | p_oodc | p_ood_re |
|--------|-------|----------|------|-------|--------|----------|
| **walldist-log** | 3 | **0.3893 ± 0.003** | **12.97 ± 0.32** | 34.03 ± 0.32 | 8.63 ± 0.21 | **24.57 ± 0.25** |
| walldist-raw | 2 | 0.3951 ± 0.001 | 13.50 ± 0.42 | 34.35 ± 0.07 | 8.75 ± 0.35 | 24.75 ± 0.35 |
| walldist-log lr=3e-4 | 1 | 0.3940 | 13.7 | 35.3 | 8.7 | 24.7 |
| baseline (this run) | 2 | 0.4083 ± 0.006 | 13.30 ± 0.00 | 34.25 ± 1.06 | 8.55 ± 0.35 | 24.90 ± 0.00 |
| **baseline (PR body, 8-seed)** | 8 | **0.404 ± 0.004** | **13.33 ± 0.58** | **33.57 ± 0.44** | **8.37 ± 0.22** | **24.58 ± 0.13** |

### Delta vs 8-seed baseline

| Config | Δ val/loss | Δ p_in | Δ p_tan | Δ p_oodc | Δ p_ood_re |
|--------|-----------|--------|---------|----------|------------|
| walldist-log | **-3.6%** ✅ | **-2.7%** ✅ | +1.4% | +3.1% | -0.04% |
| walldist-raw | **-2.2%** ✅ | +1.3% | +2.3% | +4.5% | +0.7% |

### Peak memory
- Walldist: 36.3 GB (baseline: 36.2 GB) — negligible overhead

### Commands used

Base command (all runs):
```bash
cd cfd_tandemfoil && python train.py --agent alphonse --wandb_group phase5/walldist \
  --field_decoder --adaln_output --use_lion --lr 2e-4 --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --weight_decay 5e-5 --cosine_T_max 180 --disable_pcgrad --pressure_first --pressure_deep
```
Walldist variants add: `--walldist` (log version) or `--walldist --walldist_log False` (raw version).

### What happened

**Walldist-log is a clear improvement on val/loss**, reducing it from 0.404 to 0.389 (3.6% vs 8-seed baseline, 3.7σ beyond baseline std). This is consistent across all 3 seeds (0.386–0.392), well outside the baseline noise band.

Breaking it down by component:
- **val_in_dist loss**: 0.262 avg vs 0.282 baseline → 7.1% improvement (biggest win)
- **val_ood_re loss**: 0.153 avg vs 0.174 baseline → 12.1% improvement
- **val_ood_cond loss**: 0.165 avg vs 0.175 baseline → 5.7% improvement
- **val_tandem loss**: 0.978 avg vs 0.986 baseline → 0.8% improvement (modest)

Surface MAE picture is mixed:
- **p_in improved**: 12.97 vs 13.33 (-2.7%)
- **p_tan essentially flat**: 34.03 vs 33.57 (+1.4%, within 1σ of baseline)
- **p_oodc slightly worse**: 8.63 vs 8.37 (+3.1%, 1.2σ above baseline mean)
- **p_ood_re flat**: 24.57 vs 24.58

**Volume MAE improved significantly** (e.g., vol_Ux: 0.38 vs 0.56 for best walldist seed — a 32% reduction). The wall distance feature helps the model understand the spatial relationship to boundaries, which benefits volume predictions throughout the domain.

**Walldist-raw also improved val/loss** (0.395 vs 0.404, -2.2%) but less than walldist-log, confirming the log transform helps compress the range.

**Higher lr=3e-4 did not help** (0.394 vs 0.389 for walldist-log at lr=2e-4).

### Caveats

**Epoch count difference**: Walldist runs got 155–158 epochs vs baseline 145. This happened because baselines (GPUs 6-7) finished first (~02:10 UTC), freeing CPU resources for the remaining walldist runs (GPUs 0-5), which were still training until ~03:30 UTC. The extra ~10 epochs are at very low LR (cosine schedule: LR ≈ 7e-6 at epoch 155 vs 1.8e-5 at epoch 145), so direct training improvement is minimal, but EMA averaging benefits from ~10 additional averaging steps. This is a mild confounding factor.

### Suggested follow-ups

1. **Multi-seed confirmation at matched epoch count**: Run walldist-log with a lower `cosine_T_max` to match the ~145 epoch count of baselines and confirm the improvement is from the feature, not extra EMA epochs.
2. **Walldist + foil2_surface_weight**: Since walldist improves in-dist and volume MAE but doesn't help tandem transfer much, combining with a mild foil-2 surface weight boost (sw=3) might complement each other.
3. **Walldist as input to a separate learned embedding**: Instead of concatenating raw walldist to input, use it as an auxiliary feature in a small MLP to modulate the hidden state (similar to the time embedding in diffusion models).
